src/charactersheet/viewmodels/character/spells/index.js

```diff
@@ -1,294 +1,160 @@
+            'range asc': { field: 'range', direction: 'asc'},
+            'range desc': { field: 'range', direction: 'desc'}
+        };
+    }
+
+    highestLevelSpellSlot = ko.pureComputed(() => {
+        const availableSlots = filter(this.spellSlots(), (spellSlot)=> (spellSlot.used() < spellSlot.max()));
+        const highestSlot = maxBy(availableSlots, (spellSlot) => (spellSlot.level()));
+        if (highestSlot) {
+            return highestSlot.level();
+        }
+        return 0;
+    })
 
-    self.numberOfPrepared = ko.computed(function() {
-        var prepared = 0;
-        self.spellbook().forEach(function(spell) {
-            if (spell.prepared() === true) {
-                prepared++;
+    filteredAndSortedEntities = ko.pureComputed(() =>  {
+        let spellbook = this.entities();
+        if (this.filteredByCastable()) {
+            spellbook = this.entities().filter(this.spellIsCastable, this);
+        }
+        return SortService.sortAndFilter(spellbook, this.sort(), null);
+    }, this);
+
+    toggleSpellFilter = () => {
+        this.filteredByCastable(!this.filteredByCastable());
+    }
+
+    getDefaultSort () {
+        return this.sorts()['level asc'];
+    }
+
+    setUpSubscriptions () {
+        super.setUpSubscriptions();
+        this.subscriptions.push(Notifications.spellslot.changed.add(this.updateSpellSlots, this));
+        this.subscriptions.push(Notifications.spellslot.added.add(this.updateSpellSlots, this));
+        this.subscriptions.push(Notifications.spellslot.deleted.add(this.updateSpellSlots, this));
+        this.subscriptions.push(Notifications.spellstats.changed.add(this.updateSpellStats, this));
+    }
+
+    updateSpellStats = async (stats) => {
+        this.spellStats().importValues(stats.exportValues());
+    }
+
+    updateSpellSlots = async (slot) => {
+        const response = await SpellSlot.ps.list({ coreUuid: this.coreKey });
+        this.spellSlots(response.objects);
+    }
+
+    spellIsCastable = (spell) => {
+        if (spell.level() === 0) {
+            return true;
+        } else if (spell.level() > this.highestLevelSpellSlot()) {
+            return false;
+        } else if (spell.level() === 0) {
+            return true;
+        } else if ( spell.alwaysPrepared() === true) {
+            return true;
+        } else if (spell.prepared() === true) {
+            return true;
+        }
+        return false;
+    }
+
+    spellTypeLabel = (spell) => {
+        switch(spell.type()) {
+        case 'Savings Throw': {
+            return (
+              `Save vs <strong>${spell.spellSaveAttribute()}</strong>` +
+              ` <span class="small">DC:<strong>${this.spellStats().spellSaveDc()}</strong></span>`);
+        }
+        case 'Melee Spell Attack': {
+            const spellAttackLabel = this.spellStats().spellAttackBonus();
+            let sign = ' +';
+            if (parseInt(spellAttackLabel) < 0 ) {
```

Comment:
unneeded cast

Reply:
lol. as you may be able to tell, I've been burned by js numbers. I'll remove.